src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/BQSRSpark.java

```diff
@@ -0,0 +1,102 @@
+import org.broadinstitute.hellbender.exceptions.UserException;
+import org.broadinstitute.hellbender.tools.ApplyBQSRArgumentCollection;
+import org.broadinstitute.hellbender.tools.spark.transforms.ApplyBQSRSparkFn;
+import org.broadinstitute.hellbender.tools.spark.transforms.BaseRecalibratorSparkFn;
+import org.broadinstitute.hellbender.tools.walkers.bqsr.BaseRecalibrator;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.ReadsWriteFormat;
+import org.broadinstitute.hellbender.utils.recalibration.BaseRecalibrationEngine;
+import org.broadinstitute.hellbender.utils.recalibration.RecalibrationArgumentCollection;
+import org.broadinstitute.hellbender.utils.recalibration.RecalibrationReport;
+import org.broadinstitute.hellbender.utils.variant.Variant;
+
+import java.io.IOException;
+import java.util.List;
+@CommandLineProgramProperties(
+        summary = "This tools performs 2 steps of BQSR - crection of recalibration tables and rewriting of the bam, without writing the tables to disk. ",
+        oneLineSummary = "Both steps of BQSR",
+        usageExample = "BQSRSpark -I in.bam --BQSRKnownVariants in.vcf -O out.bam",
+        programGroup = SparkProgramGroup.class
+)
+/**
+ * BQSR. The final result is analysis-ready reads.
+ */
+public final class BQSRSpark extends GATKSparkTool {
+    private static final long serialVersionUID = 1L;
+
+    @Override
+    public boolean requiresReads() { return true; }
+
+    @Override
+    public boolean requiresReference() { return true; }
+
+    @Argument(doc = "the known variants", shortName = "BQSRKnownVariants", fullName = "baseRecalibrationKnownVariants", optional = true)
+    protected List<String> baseRecalibrationKnownVariants;
+
+    @Argument(doc = "the output bam", shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME,
+            fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, optional = false)
+    protected String output;
+
+    @Argument(doc = "If specified, shard the output bam", shortName = "shardedOutput", fullName = "shardedOutput", optional = true)
+    private boolean shardedOutput = false;
+
+    @Argument(doc = "the join strategy for reference bases", shortName = "joinStrategy", fullName = "joinStrategy", optional = true)
+    private JoinStrategy joinStrategy = JoinStrategy.SHUFFLE;
+    
+    @Override
+    public SerializableFunction<GATKRead, SimpleInterval> getReferenceWindowFunction() {
+        return BaseRecalibrationEngine.BQSR_REFERENCE_WINDOW_FUNCTION;
+    }
+
+    @Override
+    protected void runTool(final JavaSparkContext ctx) {
+        JavaRDD<GATKRead> markedReads = getReads();
+        VariantsSparkSource variantsSparkSource = new VariantsSparkSource(ctx);
+
+        if ( baseRecalibrationKnownVariants.isEmpty() ) { // Warn the user if no dbSNP file or other variant mask was specified
```

Comment:
hmm, that's different than GATK3 and different than BaseRecalibrator. How about we make it non optional throughout?